### DISCUSSION TITLE
How to setup Tiled plugin by parcel v2?

### DISCUSSION THREAD

**Discussion by tenpaMk2** [excaliburjs/Excalibur#2370] [Q&A] [answered]
Hi, I want to use Tiled and setup Tiled-plugin by parcel v2.
But, something wrong in my code. 

I have 3 Tiled files.

1. `sample-stage.tmx`
2. `kenney-pixel-platformer.tsx`
3. `tiles_packed.png`

## my `.parcelrc` is correct?

```json
{
  "extends": "@parcel/config-default",
  "transformers": {
    "*.{jpg,png,bmp,tmx,tsx,json,wav,mp3,ogg}": [
      "@parcel/transformer-raw"
    ]
  }
}
```

It's copied by ludum-50

## my `resource.ts` is correct?

```ts
import { TiledMapResource } from "@excaliburjs/plugin-tiled";
import { ImageSource } from "excalibur";

// import mapchipPng from "./assets/roguelikeSheet_transparent.png";
import tiledMapTmx from "./assets/sample-stage.tmx";

export const Resources = {
  // mapchip: new ImageSource(mapchipPng),
  tiledmap: new TiledMapResource(tiledMapTmx),
};
```

## result

I got the error: `[Error] :  Error loading resources, things may not behave properly TypeError: Cannot set properties of undefined (setting 'firstgid')` .
And, there are no  `kenney-pixel-platformer.tsx` and no `tiles_packed.png` in `dist` directory.

Can anyone advise me?
Or, Should I use Tiled-plugin by webpack?

**Comment by ewal** [comment, marked as answer]
I've been fighting with both Parcel and Vite but ended up using Webpack when working with Tiled files. (I'm gonna give parcel another shot now that I know the basics of the problem since Webpack is kinda slow).

TLDR;
Assets will be served from a different a path when running the web server so the folder structure is important.

For example when using WebPack 5 (and I think this is true for all the build/serve tools): if you have a .tmx file with the path `./src/maps/map.tmx` it will be served as `/[some_hash].txm` from the dev server.
If you store your images and/or tilesets in a different path than your map files you will run into problems since they are referenced by a relative path.

Example:
`./src/maps/map.tmx`
might include
`../tilesets/terrain.tsx`
which in turn might include...
`../../images/terrain.png`

The web server will not know how to find these assets since the map file have a new root plus they doesn't exists in the output folder because they are referenced from within a file, ie. not imported as an asset so webpack don't know they exist.

I solved it by keeping all assets in the same folder and kept the maps and tileset files in the root directory of that folder. That way the files will never be referenced by first moving back into a non existent parent dir.
The second thing is to copy all assets that is referenced within a map or tileset, ie. not handled by webpack, to the dev server output dir.

Hope that helps :)

**Reply by tenpaMk2** [reply]
@ewal , Thanks advice!!

TLDR;
I succeed to use tiled-plugin under parcel v2!!
See my [template repository](https://github.com/tenpaMk2/excalibur-tiled-parcel2-template)

> I solved it by keeping all assets in the same folder and kept the maps and tileset files in the root directory of that folder. That way the files will never be referenced by first moving back into a non existent parent dir.

OK. I think this is important.

> The second thing is to copy all assets that is referenced within a map or tileset, ie. not handled by webpack, to the dev server output dir.

I found the [parcel-reporter-static-files-copy](https://github.com/elwin013/parcel-reporter-static-files-copy) plugin.
This works the above tasks!!
I created the template using this.

**Reply by ewal** [reply]
Nice job! I'm thinking of making a template for WebPack too (even though I'll personally switch to Parcel soon :))

**Reply by DaVince** [reply]
Thanks for this. I was struggling with getting a map loaded too, and your solution worked.

I have found a bug relating to this: the very moment I try putting all the map files and related files in a subdirectory, Excalibur cannot find the related files anymore. It'll load the map and then fail to find my tileset and image files. It seems that Excalibur **always looks in the root URL of the server for any files related to the Tiled map**.

I "solved" it by putting all my map resources straight into the `static` directory but I would really have preferred to keep them all in a `maps` subdirectory.

Edit: @eonarheim I should probably report this as a bug, but now that it works I'd rather work on my game. Notifying you here in case I forget to create a proper bug issue later. :sweat_smile:

**Reply by chrisMYchen** [reply]
For any future Tiled <> Excalibur warriors, https://github.com/excaliburjs/excalibur-tiled#handling-tiled-paths mentions how you can handle other file hierarchies (other than everything in root). 

```
// Create a new TiledResource loadable
var map = new ex.Plugin.Tiled.TiledMapResource("map.tmx");

map.convertPath = function (originPath, relativePath) {
   return "/assets/tx/" + path;
}
```

**Reply by tenpaMk2** [reply]
Thanks for the information!!
I understood that `convertPath` is not needed when you **work** in the following structure as it is.

```
assets/
  - maps/map.json
  - tx/map.png
  - tsx/map.tsx
```

This is useful function for the people who have different work-structure and game-structure.

**Reply by DaVince** [reply]
@chrisMYchen Thing is, I've seen that documentation and tried following it, but I _still_ don't understand what's going on there. Even in your example you provide a different path but it still doesn't click how that works. If I return a different directory, I guess it will load everything from that directory? Does that include the tmx itself?

**Reply by chrisMYchen** [reply]
@DaVince 
The copy-pasted docs have a confusing function definition. LMK if this helps! Otherwise if you want to open a new discussion with a link to your repo I'm happy to take a look too. (This thread is getting a bit long.)

If you look in your `.tmx` file, you'll see the path(s) for your tileset/resources you're using that Tiled uses to locate your tiles. All these references are probably at the top level too based on your description earlier.

If you have a `/static` directory but I want to have everything in a maps subdirectory, what you can do is use `convertPath` to change those references to include `/maps`.

```
import * as path from "path"

map.convertPath = function (pathToMapFile, relativePathToResource) {
     // here you should modify this line to return the path that matches that resource in your parcel build (`/maps/`)
   return path.join('/maps/, relativePathToResource)
}
```

Didn't test any of this - but I think these function parameter names make things a bit clearer!
